errors: narrow error types accepted by `HistoryListener`

muzarski opened the pull request:
Ref: https://github.com/scylladb/scylla-rust-driver/issues/519

## New error types and variants

### RequestError (reintroduced)
Error type returned by `run_request_speculative_fiber`. It represents either a definite request failure - last attempt error, connection pool error, or an empty plan error.

### Replaced QueryError::RequestTimeout
This error variant was not good for 2 reasons:
1. it contains stringified error message
2. it was used for both user requests timeouts, and schema agreement timeouts
I decoupled this variant into two separate variants that provide more context.

### RequestError::RequestTimeout
After refactoring the timeout errors in `QueryError`, I added the corresponding variant to `RequestError` (one introduced in this PR) as well.

## HistoryListener
### Narrowed error types passed to log_query_error and log_attempt_error.
`log_query_error` (or rather `log_request_error` after rename) now accepts `RequestError`. `log_attempt_error` now accepts `RequestAttemptError` (representing an error of a single attempt).

### Replaced "query" mentions with "request"
I did the replacement in trait methods, docstrings, tests etc. I tried to break the changes into multiple commits.

## Execution errors summary
Currently, in session layer we can distinguish 3 kind of errors:
- `RequestAttemptError` - a failure of a single attempt of request execution. It is used in public API in `RetrySession::decide_should_retry` and `HistoryListener::log_attempt_error`
- `RequestError` - a definite failure of request execution, after all (speculative) retries etc. In addition to `RequestAttemptError` it contains information about the empty plan and request timeout. It also contains `ConnectionPoolError` variant. This variant is not included in `RequestAttemptError`, because we don't want to log such error in execution history - there is a comment suggesting that (is it still true, though?). `RequestError` is used in public API in `HistoryListener::log_request_error`.
- `QueryError` - a top-level error returned by `Session` methods. It holds some additional information - e.g. about the failure of request preparation - values serialization etc.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~[ ] I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- [x] I added appropriate `Fixes:` annotations to PR description.

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -870,6 +870,38 @@ pub enum CqlEventHandlingError {
     SendError,
 }
 
+/// Failed to execute a retriable user request.
+#[derive(Error, Debug, Clone)]
+#[non_exhaustive]
+pub enum RetriableRequestError {
```
I also dislike this name. I just couldn't come up with anything better. My only idea is to rename `UserRequestError` -> `[User]RequestAttemptError` (I like this name, because it suggests that it's an error of a single attempt). Then we could use `UserRequestError` name instead of `RetriableRequestError`.

What do you think of `TimeoutableRequestError` name (introduced in this PR)? I'm not sure if you reached the commit that introduces it - this enum simply represents either a request timeout, or a request failure (current `RetriableRequestError`). I wonder if there is a better name for this one as well.

wprzytula replied:
> My only idea is to rename `UserRequestError` -> `[User]RequestAttemptError` (I like this name, because it suggests that it's an error of a single attempt). Then we could use `UserRequestError` name instead of `RetriableRequestError`.

Sounds good, let's do that.

> What do you think of `TimeoutableRequestError` name (introduced in this PR)? I'm not sure if you reached the commit that introduces it - this enum simply represents either a request timeout, or a request failure (current `RetriableRequestError`). I wonder if there is a better name for this one as well.

`MaybeTimeoutedRequestError` is one idea that I've had, but it's not ideal, too.

muzarski commented:
v2:
- rebased on most recent https://github.com/scylladb/scylla-rust-driver/pull/1157
- renamed `UserRequestError` -> `UserRequestAttemptError` and `RetriableRequestError` -> `UserRequestError`

muzarski commented:
Rebased on latest https://github.com/scylladb/scylla-rust-driver/pull/1157 (with modules refactor)

muzarski commented:
v2.1: added a regression test for the scenario where the driver would unnecessarily wait for `retry_interval` and do the speculative retry in case of empty plan. The test's logic is explained in the (first) commit.

wprzytula commented on scylla/tests/integration/retries.rs:
```diff
@@ -232,3 +238,75 @@ async fn speculative_execution_panic_regression_test() {
+#[tokio::test]
+#[ntest::timeout(5000)]
+#[cfg(not(scylla_cloud_tests))]
+async fn speculative_execution_does_not_retry_on_emtpy_plan_regression_test() {
```
:pick: `emtpy`

wprzytula commented on scylla/tests/integration/retries.rs:
```diff
@@ -232,3 +238,75 @@ async fn speculative_execution_panic_regression_test() {
     }
 }
+
+/// This is a regression test.
+/// We check that speculative execution does not retry when the load balancing policy returns an empty plan.
+/// The test timeout is set to 5 seconds, so the speculative execution policy with a retry interval 6s cannot retry.
+#[tokio::test]
+#[ntest::timeout(5000)]
```
:question: :thought_balloon: Can we be sure that 5 seconds is enough for running the whole test in the CI? What's the usual timeout of such session-building tests?

Lorak-mmk reviewed (changes requested):
I'll review whole PR soon, but I'm posting this comment separately because it may be important. Correct me if I'm wrong, but commit "spec_exec: flatten return type of run_request_speculative_fiber" seems to misunderstand speculative execution logic and introduce a bug because of that. 

As you noted, the old logic ignores the empty plan error. You changed it to return immediately, which I think is not correct.

Consider the following scenario:
- We start some speculative fibers, some of them are still executing (e.g. sent a request and waiting for an answer). They exhausted the plan (note that the plan is shared between fibers!)
- Timer goes off, we start another fiber
- This new fiber gets an empty plan, so returns immediately
- We should ignore this error, because not all existing fibers finished yet. You new logic would return immediately.

There is  one more issue: let's say that in the above scenario the "empty plan" fiber was actually the last fiber (and so we should return an error now, because the execution definitely failed). The error we should return is not the "empty plan" error from this fiber, but the previous error. This is because empty plan error is just a symptom of all other executions failing.
This is what this code did in the old logic:
```rust
 return last_error.unwrap_or({
      Err(EMPTY_PLAN_ERROR)
  });
```
This caused us to return EMPTY_PLAN_ERROR only if last_error was None. If we finished all executions (`async_tasks.is_empty() && retries_remaining == 0`), and the error is None, that means no attempt returned an error, so there was no attempt at all, so the plan was empty. This is the only scenario where we want to return the empty plan error.

muzarski commented:
> I'll review whole PR soon, but I'm posting this comment separately because it may be important. Correct me if I'm wrong, but commit "spec_exec: flatten return type of run_request_speculative_fiber" seems to misunderstand speculative execution logic and introduce a bug because of that.
> 
> As you noted, the old logic ignores the empty plan error. You changed it to return immediately, which I think is not correct.
> 
> Consider the following scenario:
> 
>     * We start some speculative fibers, some of them are still executing (e.g. sent a request and waiting for an answer). They exhausted the plan (note that the plan is shared between fibers!)
> 
>     * Timer goes off, we start another fiber
> 
>     * This new fiber gets an empty plan, so returns immediately
> 
>     * We should ignore this error, because not all existing fibers finished yet. You new logic would return immediately.
> 
>
> There is one more issue: let's say that in the above scenario the "empty plan" fiber was actually the last fiber (and so we should return an error now, because the execution definitely failed). The error we should return is not the "empty plan" error from this fiber, but the previous error. This is because empty plan error is just a symptom of all other executions failing. This is what this code did in the old logic:
> 
> ```rust
>  return last_error.unwrap_or({
>       Err(EMPTY_PLAN_ERROR)
>   });
> ```
> 
> This caused us to return EMPTY_PLAN_ERROR only if last_error was None. If we finished all executions (`async_tasks.is_empty() && retries_remaining == 0`), and the error is None, that means no attempt returned an error, so there was no attempt at all, so the plan was empty. This is the only scenario where we want to return the empty plan error.

You are 100% correct, this is a bug. Thanks for spotting it. After giving it a thought, I think that I should revert this change - i.e. `Option<Result<...>>` is a good choice for a return type of `run_request_speculative_fiber`.

I also wonder if we should handle scenarios such as presented in the regression test. In other words, not to unnecessarily wait for a retry if the following conditions are true:
- finished task returned `None` (plan was empty/exhausted)
- there are no other running fibers - for checking this, we can introduce additional variable which would store the value of currently running fibers

Do you think that it makes sense to prevent from such scenarios? Or is it such rare scenario that it's not worth to introduce additional noise to the function (new variable, additional branching etc.)

muzarski commented on scylla/src/observability/history.rs:
```diff
@@ -619,7 +619,7 @@ mod tests {
+            &unexpected_response(CqlResponseKind::Ready),
             &RetryDecision::RetrySameNode(Some(Consistency::Quorum)),
         );
 
```
This is a perfect example of why `QueryError` was "too broad" for `log_attempt_error`. QueryError::TimeoutError could never be passed to this method, because it's an error constructed in the higher layer of execution (https://github.com/muzarski/scylla-rust-driver/blob/main/scylla/src/client/session.rs#L1974-L1987) which is done after we log all attempt errors. Since we narrowed the error type passed to `log_attempt_error`, I needed to adjust the test accordingly (there is no `TimeoutError` variant in `RequestAttemptError`).

muzarski commented on scylla/src/errors.rs:
```diff
@@ -904,6 +904,13 @@ pub enum RequestError {
     #[error("No connections in the pool: {0}")]
     ConnectionPoolError(#[from] ConnectionPoolError),
 
+    /// Failed to run a request within a provided client timeout.
+    #[error(
+            "Request execution exceeded a client timeout {}ms",
+            std::time::Duration::as_millis(.0)
+        )]
+    RequestTimeout(std::time::Duration),
+
```
It is constructed in the next commit here https://github.com/scylladb/scylla-rust-driver/pull/1159/commits/4bae738659e81d74be0a4cc22c35624002594e17#diff-8dcf6c908c350e5e0977344268a9ba9095466e548365d416b436c938a83e49d2R1980-R1982 (I hope that this link works as expected)

Lorak-mmk commented:
> 
> You are 100% correct, this is a bug. Thanks for spotting it. After giving it a thought, I think that I should revert this change - i.e. `Option<Result<...>>` is a good choice for a return type of `run_request_speculative_fiber`.
> 
> I also wonder if we should handle scenarios such as presented in the regression test. In other words, not to unnecessarily wait for a retry if the following conditions are true:
> 
>     * finished task returned `None` (plan was empty/exhausted)
> 
>     * there are no other running fibers - for checking this, we can introduce additional variable which would store the value of currently running fibers
> 
> 
> Do you think that it makes sense to prevent from such scenarios? Or is it such rare scenario that it's not worth to introduce additional noise to the function (new variable, additional branching etc.)

IIUC this would mostly help in a scenario where the plan was already exhausted, but there are still some speculative execution left to start ("retries_remaining > 0").
We could of course do this, but I'm a bit worried about the possibility of bugs, both now and in the future.
While this piece of code doesn't really look complicated, it already had a bug that I fixed in the past: https://github.com/scylladb/scylla-rust-driver/pull/1086 ,
and now you nearly introduced another one. For me this is an indication that this piece of logic is more complicated and error-prone than it seems at first glance, so making it more complicated may backfire at us.

Maybe you have an idea on how to add this functionality while also rewriting this logic to be less dangerous?

muzarski commented on scylla/src/errors.rs:
```diff
@@ -871,7 +871,47 @@ pub enum CqlEventHandlingError {
+        )]
+    EmptyPlan,
+
+    /// Selected node's connection pool is in invalid state.
+    #[error("No connections in the pool: {0}")]
+    ConnectionPoolError(#[from] ConnectionPoolError),
+
```
This is a great question. I wanted to put it under `RequestAttemptError`, but then I saw the comment: https://github.com/muzarski/scylla-rust-driver/blob/main/scylla/src/client/session.rs#L2033-L2035. It says that we should not log such error in metrics. By metrics, I understand the request execution history. If I introduced the corresponding variant in `RequestAttemptError`, we once again would be passing a "too broad" error type to `HistoryListener::log_attempt_error` - variant corresponding to `ConnectionPoolError` would never be constructed.

muzarski commented:
> IIUC this would mostly help in a scenario where the plan was already exhausted, but there are still some speculative execution left to start ("retries_remaining > 0").

Exactly.

> We could of course do this, but I'm a bit worried about the possibility of bugs, both now and in the future. While this piece of code doesn't really look complicated, it already had a bug that I fixed in the past: #1086 , and now you nearly introduced another one. For me this is an indication that this piece of logic is more complicated and error-prone than it seems at first glance, so making it more complicated may backfire at us.

I 100% agree.

> Maybe you have an idea on how to add this functionality while also rewriting this logic to be less dangerous?

I can think about it, but this is probably out of scope of this PR. I'll open an issue and I believe this could even be addressed in post-1.0.0 driver.

Lorak-mmk commented:
> > We could of course do this, but I'm a bit worried about the possibility of bugs, both now and in the future. While this piece of code doesn't really look complicated, it already had a bug that I fixed in the past: #1086 , and now you nearly introduced another one. For me this is an indication that this piece of logic is more complicated and error-prone than it seems at first glance, so making it more complicated may backfire at us.
> 
> I 100% agree.
> 

Random thought: I wonder (not for the first time) if Rust makes us developers more "lazy" because of how many problems it eliminates automatically.
By "lazy" I mean less careful and thus more likely to not notice some bugs, than developers using other languages.

Lorak-mmk commented:
One more thing: could you provide a very short overview in description on request-error related types after the PR?
By that I mean what types (RequestError? QueryError? RequestAttemptError? maybe some others I don't know) exist and what is their purpose.

muzarski commented:
v3:
- rebased on main
- dropped the first commit that would change the return type of `run_request_speculative_fiber` from `Option<Result<_>>` to `Result<_>`
- addressed minor issues from @Lorak-mmk comments
- updated cover letter - added execution error types summary

Lorak-mmk reviewed (approved):
Looks fine now. I'm not fully convinced about the division between QueryError (btw will it be renamed?), RequestError and RequestAttemptError. But that is something that may be better addressed by discussion at the office some day.

muzarski commented:
> Looks fine now. I'm not fully convinced about the division between QueryError (btw will it be renamed?)

Yes, I'd like to rename it. We discussed it with @wprzytula and figured out that `ExecutionError` would fit here.
